kernel/src/table_changes/mod.rs

```diff
@@ -0,0 +1,191 @@
+//! In-memory representation of a change data feed table.
+
+use std::{collections::HashMap, sync::Arc};
+
+use crate::{
+    actions::{Metadata, Protocol},
+    features::ColumnMappingMode,
+    log_segment::{LogSegment, LogSegmentBuilder},
+    path::AsUrl,
+    scan::{get_state_info, state::DvInfo, ColumnType},
+    schema::{Schema, SchemaRef, StructType},
+    snapshot::Snapshot,
+    DeltaResult, Engine, EngineData, Error, ExpressionRef, Version,
+};
+use url::Url;
+
+pub type TableChangesScanData = (Box<dyn EngineData>, Vec<bool>, Arc<HashMap<String, DvInfo>>);
+
+static CDF_ENABLE_FLAG: &str = "delta.enableChangeDataFeed";
+
+#[derive(Debug)]
+pub struct TableChanges {
+    pub snapshot: Snapshot,
+    #[allow(unused)]
+    pub(crate) log_segment: LogSegment,
+    pub schema: Schema,
+    pub version: Version,
+    pub metadata: Metadata,
+    pub protocol: Protocol,
+    pub(crate) column_mapping_mode: ColumnMappingMode,
+    pub table_root: Url,
+}
+
+impl TableChanges {
+    pub fn try_new(
+        table_root: Url,
+        engine: &dyn Engine,
+        start_version: Version,
+        end_version: Option<Version>,
+    ) -> DeltaResult<Self> {
+        let start_snapshot =
+            Snapshot::try_new(table_root.as_url().clone(), engine, Some(start_version))?;
+        let end_snapshot = Snapshot::try_new(table_root.as_url().clone(), engine, end_version)?;
+        println!("Start snapshot: {:?}", start_snapshot);
+        println!("End snapshot: {:?}", end_snapshot);
+
+        let start_flag = start_snapshot.metadata().configuration.get(CDF_ENABLE_FLAG);
+        let end_flag = end_snapshot.metadata().configuration.get(CDF_ENABLE_FLAG);
+
+        // Verify CDF is enabled at the beginning and end of the interval
+        let is_valid_flag = |flag_res: Option<&String>| flag_res.is_some_and(|val| val == "true");
+        if !is_valid_flag(start_flag) || !is_valid_flag(end_flag) {
+            return Err(Error::TableChangesDisabled(start_version, end_version));
+        }
+
+        println!("Validated flags");
+
+        // Get a log segment for the CDF range
+        let fs_client = engine.get_file_system_client();
+        let mut builder = LogSegmentBuilder::new(fs_client, &table_root);
+        builder = builder.with_start_version(start_version);
+        if let Some(end_version) = end_version {
+            builder = builder.with_end_version(end_version);
+        }
+        builder = builder
+            .with_no_checkpoint_files()
+            .with_in_order_commit_files();
+        let log_segment = builder.build()?;
+
+        println!("Built log segment");
+        Ok(TableChanges {
+            snapshot: start_snapshot,
+            log_segment,
+            schema: end_snapshot.schema().clone(),
```

Comment:
If the snapshot keeps a lot of the struct members anyways, why not just keep the ending snapshot instead of all the clones?

Reply:
Godo point, I was trying to avoid holding the `end_snapshot`'s `LogSegment`, but perhaps we're fine with carrying all that with us.